p2p/net/reuseport/dialer.go

```diff
@@ -0,0 +1,91 @@
+package reuseport
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+
+	"github.com/libp2p/go-netroute"
+)
+
+type dialer struct {
+	specific    []*net.TCPAddr
+	loopback    []*net.TCPAddr
+	unspecified []*net.TCPAddr
+}
+
+func (d *dialer) Dial(network, addr string) (net.Conn, error) {
+	return d.DialContext(context.Background(), network, addr)
+}
+
+func randAddr(addrs []*net.TCPAddr) *net.TCPAddr {
+	if len(addrs) > 0 {
+		return addrs[rand.Intn(len(addrs))]
+	}
+	return nil
+}
+
+// DialContext dials a target addr.
+// Dialing preference is
+// * If there is a listener on the local interface the OS expects to use to route towards addr, use that.
+// * If there is a listener on a loopback address, addr is loopback, use that.
+// * If there is a listener on an undefined address (0.0.0.0 or ::), use that.
+// * Otherwise, dial with a random port on the unspecified address.
+func (d *dialer) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	if len(d.specific) > 0 || len(d.loopback) > 0 {
```

Comment:
Because users _usually_ just listen on the all address, and we can avoid double-resolving the TCP addr in that case.

Reply:
Can you add this as a comment? Also, I'm not sure I understand it. If the user listens on all the addresses, which one will be `> 0`?